presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestApplyTableScanRedirection.java

```diff
@@ -116,6 +119,29 @@ public void doesNotFireIfNoTableScan()
         }
     }
 
+    @Test(expectedExceptions = PrestoException.class, expectedExceptionsMessageRegExp = "Redirected type \\(bigint\\) for column destination_col_c does not match source type \\(varchar\\)")
+    public void testMismatchedTypes()
+    {
+        try (RuleTester ruleTester = defaultRuleTester()) {
+            // make the mock connector return a table scan on different table
+            ApplyTableScanRedirect applyTableScanRedirect = getMockApplyRedirect(
+                    ImmutableMap.of(sourceColumnHandleA, destinationColumnNameC));
+            MockConnectorFactory mockFactory = createMockFactory(Optional.of(applyTableScanRedirect));
+
+            ruleTester.getQueryRunner().createCatalog(MOCK_CATALOG, mockFactory, ImmutableMap.of());
+
+            ruleTester.assertThat(new ApplyTableScanRedirection(ruleTester.getMetadata()))
+                    .on(p -> {
+                        Symbol column = p.symbol(sourceColumnNameA, VARCHAR);
+                        return p.tableScan(TEST_TABLE_HANDLE,
+                                ImmutableList.of(column),
+                                ImmutableMap.of(column, sourceColumnHandleA));
+                    })
+                    .withSession(MOCK_SESSION)
+                    .matches(anyTree());
```

Comment:
Rule tested is not really leveraged much here, because we're not testing the rule application, but rather a failure internally.

Maybe it would make sense to add `.fails(Consumer<Exception> exceptionAssertion)` method to Rule tester, next to `matches()`?
Altenratively, maybe you do not need rule tester at all, you can just run a query on the query runner after all.